### PR TITLE
Beamline creation readme

### PR DESCRIPTION
**Pull request by DiamondJoseph**
Addresses #321 (hopefully in time to assist creation of i20-1 beamline, help get i22 merged)

### Instructions to reviewer on how to test:
1. Follow through the readme whilst creating a new beamline
2. Confirm that the created beamline passes all tests, creates sensible devices

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://github.com/DiamondLightSource/dodal/wiki/Device-Standards)

### CODE REVIEW

**Comment by DiamondJoseph** [comment]
This change does enforce that Only device factory functions are defined in the beamlines package, and also enforces that beamlines either utilise device_instantiation or some other form of ensuring their methods return a singleton.

Test coverage of new beamlines is already above the required patchset percentage just by inclusion in test_device_instantiation, which should reduce the barrier to entry of getting new beamlines added.

File: docs/user/tutorials/get_started.rst [hunk 1]
```diff
@@ -8,6 +8,75 @@ The Purpose of Dodal
 
 If you have some code that you think would be useful to be added please do so!
 
+
+Creating a new beamline
```
[the remark this answers is not in the document]
Comment (DiamondJoseph) [reply docs/user/tutorials/get_started.rst]: Done

File: tests/beamlines/unit_tests/test_i24.py [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
+from dodal.utils import make_all_devices
 
-with patch.dict("os.environ", {"BEAMLINE": "i24"}, clear=True):
-    from dodal.beamlines import beamline_utils, i24
-    from dodal.utils import make_all_devices
 
-
-def test_device_creation():
-    devices = make_all_devices(i24, fake_with_ophyd_sim=True)
+@patch.object(dodal.utils, "get_beamline_name", return_value="i24")
```
Comment (dperl-dls) [inline tests/beamlines/unit_tests/test_i24.py:11]: I prefer the solution in https://github.com/DiamondLightSource/dodal/pull/330#pullrequestreview-1877164642 - I know it's just a test, but reloading modules and deleting from sys feels extra messy to me - and it's not really necessary - all that's needed is to set the `beamline_utils.BL` and clear the devices like I did here https://github.com/DiamondLightSource/dodal/commit/05ab615825f85caef0d1cdb64ef7d3f512483e40

Comment (DiamondJoseph) [reply tests/beamlines/unit_tests/test_i24.py:11]: Now that 330 is merged I'm happy to just rebase on top of main again and discard this. Any thoughts on the rest of the changes?

Comment (dperl-dls) [reply tests/beamlines/unit_tests/test_i24.py:11]: Yes, I think this is great! it'll be very helpful I'm sure. It looks detailed enough for most developers to me, hopefully that's right. But I think you'd get better feedback from someone who didn't write the thing in the first place, haha

**Review by dperl-dls — APPROVED** [review]
Thanks for this! I see nothing objectionable here so I think we can improve it later if we can get some feedback from people who look at it with fresh eyes

File: tests/beamlines/unit_tests/test_i24.py (hunk 2 above)
Comment (DiamondJoseph) [reply tests/beamlines/unit_tests/test_i24.py]: Removed the i24 test changes from the tree

Comment (DiamondJoseph) [reply tests/beamlines/unit_tests/test_i24.py]: But now i24 is failing again

Comment (dperl-dls) [reply tests/beamlines/unit_tests/test_i24.py]: I'm looking into it, I think it's not relevant for this PR, you can merge it anyway

Comment (dperl-dls) [reply tests/beamlines/unit_tests/test_i24.py]: https://github.com/DiamondLightSource/dodal/issues/331

**Pull request merged**